pkg/config/handlerConfigurer.go

```diff
@@ -0,0 +1,191 @@
+	}
+}
+
+func newInstancesByTemplate() instancesByTemplate {
+	return instancesByTemplate{make(map[string][]string)}
+}
+
+// ConfigureHandlers identifies and invokes all the type configuration (per template) that needs
+// to be done on a handler.
+func ConfigureHandlers(actions []*pb.Action, constructors map[string]*pb.Constructor,
+	handlers map[string]*HandlerBuilderInfo, tmplRepo template.Repository, expr expr.TypeChecker, df expr.AttributeDescriptorFinder) error {
+	// Steps
+	// 1. For each handler, based on the actions it is referenced from, we first group all the
+	// constructors based on template kind. This results into something like
+	// map[handlerName]map[TemplateName][][Constructor.InstanceName].
+	// 2. We then infer all the types for all the known
+	// constructors.
+	// 3. Using data from #1 and #2, for each handler and for each template within it, we call configure*TemplateName*
+	// with all the inferred types for all the instanceNames that belong to handler-template group.
+	configurer := handlerConfigurer{tmplRepo: tmplRepo, typeChecker: expr, attrDescFinder: df}
+
+	iTypes, err := configurer.inferTypes(constructors)
+	if err != nil {
+		return err
+	}
+	grpHandlers, err := configurer.groupHandlerInstancesByTemplate(actions, constructors, handlers)
+	if err != nil {
+		return err
+	}
+
+	return configurer.dispatchTypesToHandlers(iTypes, grpHandlers, handlers)
+}
+
+func (h *handlerConfigurer) dispatchTypesToHandlers(infrdTypes map[string]proto.Message,
+	instsByTmpls map[string]instancesByTemplate, handlers map[string]*HandlerBuilderInfo) error {
+	for hName, instsByTmpl := range instsByTmpls {
+		hb, found := handlers[hName]
+		if !found {
+			// This should not happen, since the dispatchTypesToHandlers should get called after all verifications.
+			panic(fmt.Errorf("handler %s is not registered. This code should be called after config has been "+
+				"validated", hName))
+		}
+
+		for tmplName, insts := range instsByTmpl.instancesNamesByTemplate {
+			ti, found := h.tmplRepo.GetTemplateInfo(tmplName)
+			if !found {
+				// This should not happen, since the dispatchTypesToHandlers should get called after all verifications.
+				panic(fmt.Errorf("template %s is not registered. This code should be called after config has "+
+					"been validated", tmplName))
+			}
+
+			typesToConfigure := make(map[string]proto.Message)
+			for _, inst := range insts {
+				v, found := infrdTypes[inst]
+				if !found {
+					// This should not happen, since the dispatchTypesToHandlers should get called after all
+					// verifications.
+					panic(fmt.Errorf("instance %s is not found in inferred types. This code should be called "+
+						"after config has been validated", inst))
+				}
+				typesToConfigure[inst] = v
+			}
+
+			// ConfigureTypeFn calls into handler's configure code which can panic. If that happens, we will
+			// remove the handler from the list of handlers to configure.
+			defer func() {
```

Comment:
Hoist this defer out of the loop: once we get a panic, we have to stop processing anyway. There's no need to defer a func for each call.